pkg/querier/frontend2/frontend2.go

```diff
@@ -0,0 +1,290 @@
+func NewFrontend2(cfg Config, log log.Logger, reg prometheus.Registerer) (*Frontend2, error) {
+	requestsCh := make(chan *frontendRequest)
+
+	schedulerWorkers, err := newFrontendSchedulerWorkers(cfg, fmt.Sprintf("%s:%d", cfg.Addr, cfg.Port), requestsCh, log)
+	if err != nil {
+		return nil, err
+	}
+
+	f := &Frontend2{
+		cfg:              cfg,
+		log:              log,
+		requestsCh:       requestsCh,
+		schedulerWorkers: schedulerWorkers,
+		requests:         newRequestsInProgress(),
+	}
+	// Randomize to avoid getting responses from queries sent before restart (which could lead to leak between tenants).
+	// This isn't perfect, but better than nothing.
+	f.lastQueryID.Store(rand.Uint64())
+
+	promauto.With(reg).NewGaugeFunc(prometheus.GaugeOpts{
+		Name: "cortex_frontend_queries_in_progress",
+		Help: "Number of queries in progress handled by this frontend.",
+	}, func() float64 {
+		return float64(f.requests.count())
+	})
+
+	promauto.With(reg).NewGaugeFunc(prometheus.GaugeOpts{
+		Name: "cortex_frontend_connected_schedulers",
+		Help: "Number of schedulers this frontend is connected to.",
+	}, func() float64 {
+		return float64(f.schedulerWorkers.getWorkersCount())
+	})
+
+	f.Service = services.NewIdleService(f.starting, f.stopping)
+	return f, nil
+}
+
+func (f *Frontend2) starting(ctx context.Context) error {
+	return errors.Wrap(services.StartAndAwaitRunning(ctx, f.schedulerWorkers), "failed to start frontend scheduler workers")
+}
+
+func (f *Frontend2) stopping(_ error) error {
+	return errors.Wrap(services.StopAndAwaitTerminated(context.Background(), f.schedulerWorkers), "failed to stop frontend scheduler workers")
+}
+
+// RoundTripGRPC round trips a proto (instead of a HTTP request).
+func (f *Frontend2) RoundTripGRPC(ctx context.Context, req *httpgrpc.HTTPRequest) (*httpgrpc.HTTPResponse, error) {
+	if s := f.State(); s != services.Running {
+		return nil, fmt.Errorf("frontend not running: %v", s)
+	}
+
+	userID, err := user.ExtractOrgID(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Propagate trace context in gRPC too - this will be ignored if using HTTP.
+	tracer, span := opentracing.GlobalTracer(), opentracing.SpanFromContext(ctx)
+	if tracer != nil && span != nil {
+		carrier := (*httpgrpcHeadersCarrier)(req)
+		if err := tracer.Inject(span.Context(), opentracing.HTTPHeaders, carrier); err != nil {
+			return nil, err
+		}
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	freq := &frontendRequest{
+		queryID: f.lastQueryID.Inc(),
+		request: req,
+		userID:  userID,
+
+		cancel: cancel,
+
+		// Buffer of 1 to ensure response or error can be written to the channel
+		// even if this goroutine goes away due to client context cancellation.
+		enqueue:  make(chan enqueueResult, 1),
+		response: make(chan *httpgrpc.HTTPResponse, 1),
+	}
+
+	f.requests.put(freq)
+	defer f.requests.delete(freq.queryID)
+
+	retries := f.cfg.WorkerConcurrency + 1 // To make sure we hit at least two different schedulers.
```

Comment:
You mean number of retries?

The worst that can possibly happen is that no worker will take this request, and request gets stuck until context is cancelled. This can happen with 0 retries (eg. if no scheduler was found).

My thinking behind this value is:

- IF there are two schedulers,
- and one is shutting down
- and request is processed by all worker-goroutines for this scheduler (= WorkerConcurrency), 

then we still hit at least one worker for other scheduler.